### PR TITLE
Handle One Account Google SignIn

### PR DESCRIPTION
**Pull request by aweng98**
- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] **Submitter**: If you're adding new automated UI tests, review the test plan with QA
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests and automated UI tests.
- [ ] **Submitter**: Squash commits and merge to develop. If adding test code, merge application code and test code at the same time.
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by aweng98** [comment]
jenkins retest

**Comment by aweng98** [comment]
jenkins retest

**Comment by aweng98** [comment]
signinspec_oneaccountsignin

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala [hunk 1]
```diff
@@ -11,24 +11,16 @@ class SignInSpec extends FreeSpec with WebBrowserSpec with UserFixtures with Mat
...
   implicit override val patienceConfig = PatienceConfig(timeout = scaled(Span(5, Seconds)), interval = scaled(Span(500, Millis)))
 
-  "A user" - {
-    "with a registered account" - {
+  "A user with a registered account" - {
 
-      "should be able to log in and out multiple times as multiple users" in {
-        val Seq(user1, user2) = UserPool.chooseStudents(2)
+      "should be able to log in and out" in {
+        // "harry": "harry.potter@quality.firecloud.org"
```
Comment (aweng98) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala]: restrict Google real SignIn to use just one user. It's painful to use x11 forwarding for Google SignIn security question.

Comment (breilly2) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala]: We need 2 users, otherwise we lose coverage of the behavior of making sure sign-out followed by sign-in works correctly. The chance that this will fail is low but the cost if it did could be very high!

**Comment by aweng98** [comment]
sorry, I changed it because it was what Jackie wanted on Friday.
xforwarding was too painful for her to do for all test users in user pool.
It is changed to use 2 users (harry and ron) in PR. and added check to
ensure sign out really worked.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala [hunk 2]
```diff
@@ -13,14 +14,27 @@ class SignInSpec extends FreeSpec with WebBrowserSpec with UserFixtures with Mat
...
-      "should be able to log in and out" in {
-        // "harry": "harry.potter@quality.firecloud.org"
-        val student = Config.Users.testUser
-        withWebDriver { implicit driver =>
-          withSignInReal(student) { listPage =>
-            eventually { listPage.readUserEmail() shouldEqual student.email }
-          }
+    "should be able to log in and out as harry.potter" in {
+      // config is harded to get "harry": "harry.potter@quality.firecloud.org"
+      val harry = Config.Users.testUser
```
Comment (jacmrob) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala]: let's not call him "harry" here, since that value can be set in the application.conf so it's not necessarily hard-coded to harry.  let's just say 'signinUser1' or something

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala [hunk 3]
```diff
@@ -13,14 +14,27 @@ class SignInSpec extends FreeSpec with WebBrowserSpec with UserFixtures with Mat
...
+          eventually { listPage.readUserEmail() shouldEqual harry.email }
         }
+        // making sure Sign Out worked
+        eventually { new SignInPage(Config.FireCloud.baseUrl).isOpen shouldEqual true }
       }
     }
+
+    "should be able to log in and out as ron.weasley" in {
+      // config is harded to get "harry": "harry.potter@quality.firecloud.org"
+      val ron = Config.Users.Students.getUserCredential("ron")
```
Comment (jacmrob) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala]: same here-- let's add another testUser to the Config -- Config.Users.testUser2, and use that

**Review by jacmrob — CHANGES_REQUESTED** [review]

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala (hunk 2 above)
Comment (jacmrob) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala]: also since @jmthibault79's changes i'm not sure Config.Users.testUser even exists anymore-- look in FireCloudConfig.scala to confirm

**Review by jacmrob — APPROVED** [review]

**Comment by jacmrob** [comment]
jenkins retest

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala [hunk 4]
```diff
@@ -67,17 +68,17 @@ class SignInPage(val baseUrl: String)(implicit webDriver: WebDriver) extends Fir
...
+      await condition (id("identifierLink").findElement.exists(_.isDisplayed)
+        || id("identifierId").findElement.exists(_.isEnabled)
+        || id("Email").findElement.exists(_.isDisplayed), 10) // One Account All of Google popup
+    } match {
       case Success(_) =>
-        val chooseAccountLink = find(id("identifierLink")).isDefined
-        if (chooseAccountLink) {
-          click on (id("identifierLink"))
-          await visible id("identifierId")
+        find(id("identifierLink")).isDefined match {
```
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala]: I don't think the use of `match` is more readable than the `if`, especially since you're not doing anything in the `false` case.

You could try something like this:
```
find(id("identifierLink")) foreach { link =>
  click on link
  await visible id("identifierId")
}
```
It reads a little funny because `find` returns only a single element in an `Option`, but `foreach` is a perfectly valid function of `Option`... like `map` but returning `Unit` instead of transforming the value.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala [hunk 5]
```diff
@@ -108,6 +115,13 @@ class GoogleSignInPopup(implicit webDriver: WebDriver) extends WebBrowser with W
     returnFromSignIn()
   }
 
+  private def oneAccountSignIn(email: String, password: String): Unit = {
+    emailField(id("email")).value =(email)
```
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala]: Unnecessary parentheses.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala [hunk 6]
```diff
@@ -108,6 +115,13 @@ class GoogleSignInPopup(implicit webDriver: WebDriver) extends WebBrowser with W
...
+    find(id("next")).get.underlying.submit()
+    pwdField(id("Passwd")).value = password
+    find(id("signIn")).get.underlying.submit()
+  }
```
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala]: Do we need a `returnFromSignIn()` here? I don't know for sure that this new sign-in flow is in a new window but my impression from the test failure screenshots I've seen is that it is.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/library/DataUseAwareSearchSpec.scala [hunk 7]
```diff
@@ -33,11 +34,11 @@ class DataUseAwareSearchSpec extends FreeSpec with WebBrowserSpec with UserFixtu
         modal.selectCheckbox("poa")
         modal.submit()
 
-        modal.isVisible shouldBe false
+        eventually { modal.isVisible shouldBe false }
```
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/library/DataUseAwareSearchSpec.scala:37]: These changes look unrelated to sign-in... but I guess that's okay if they help test reliability. Any guidance for when we should be using `eventually`?

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala [hunk 8]
```diff
@@ -11,23 +12,28 @@ class SignInSpec extends FreeSpec with WebBrowserSpec with UserFixtures with Mat
...
-        withWebDriver { implicit driver =>
-          withSignInReal(user2) { listPage =>
-            eventually { listPage.readUserEmail() shouldEqual user2.email }
-          }
+        // making sure Sign Out worked
+        eventually { new SignInPage(FireCloudConfig.FireCloud.baseUrl).isOpen shouldEqual true }
+      }
+    }
+
+    "should be able to log in and out as ron.weasley" in {
```
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala]: The original intent of this test was lost somewhere along the way. Signing-in as a 2nd user should be in the same test as the other sign-in. It verifies that the new sign-in is respected immediately after sign-out so it's important that the same browser session is used (inside the same `withWebDriver` block).

**Review by breilly2 — CHANGES_REQUESTED** [review]
Mostly small things but worth fixing.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala (hunk 4 above)
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala]: I like this

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala (hunk 6 above)
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala]: seems true. I added `returnFromSignIn` in `signIn`

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/library/DataUseAwareSearchSpec.scala (hunk 7 above)
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/library/DataUseAwareSearchSpec.scala:37]: unrelated. impulsive.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala (hunk 8 above)
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala]: done

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala [hunk 9]
```diff
@@ -104,8 +110,13 @@ class GoogleSignInPopup(implicit webDriver: WebDriver) extends WebBrowser with W
     Thread sleep 1000
     pwdField(name("password")).value = password
     pressKeys("\n")
+  }
 
-    returnFromSignIn()
```
Comment (aweng98) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala:108]: moved to `signIn` because it's needed by `oneAccountSignIn` and `normalSignIn`

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala [hunk 10]
```diff
@@ -119,15 +130,18 @@ class GoogleSignInPopup(implicit webDriver: WebDriver) extends WebBrowser with W
...
-    /*
-     * If there is still more than 1 window after 30 seconds, we most likely
-     * need to approve access to continue.
-     */
-    if (windowHandles.size > 1) {
-      click on id("submit_approve_access")
-      await condition(windowHandles.size == 1)
+    try {
+      await condition (windowHandles.size == 1)
+    } catch {
```
Comment (aweng98) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala:135]: bug in original code. timeout exception thrown if condition is not true.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala [hunk 11]
```diff
@@ -11,24 +12,29 @@ class SignInSpec extends FreeSpec with WebBrowserSpec with UserFixtures with Mat
...
-        val Seq(user1, user2) = UserPool.chooseStudents(2)
-        withWebDriver { implicit driver =>
-          withSignInReal(user1) { listPage =>
-            eventually {
-              listPage.readUserEmail() shouldEqual user1.email
-            }
-          }
+    "should be able to log in and out as different users in same browser" in {
+      val user1 = Users.Students.getUserCredential("harry")
+      withWebDriver { implicit driver =>
```
Comment (aweng98) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/SignInSpec.scala:19]: same web browser used in 2 login

**Review by breilly2 — APPROVED** [review]
:+1:
